### PR TITLE
Local Agent DNS Lookup

### PR DESCRIPTION
**Pull request by moofish32**
Not ready to merge, but continuing the discussion on #2300

I created an agent namespace. When you query agent.consul, the behavior is simply to retrieve the NodeName from config and query that name. This returns effectively the bind address, however I think if a user has overridden this with `advertise_addr` config it will return that. I'm not sure @slackpad if this is what you intended or if this interface needs to be more intelligent. @ryansch or @far-blue is this meeting your need or did I misunderstand the use case?

### CODE REVIEW

**Comment by far-blue** [comment]
Hi @moofish32 
If I understand you correctly then yes, I think this would cover my use case perfectly.

Basically, in a setup where you want to talk to the local consul agent you can simply use, in your config, "agent.consul" and the bind address is returned. Where you have the same service deployed on multiple hosts, with each host having a local consul agent or a pre-defined 'neighbourhood' agent, the config file for that service need not change to handle talking to the local agent.

I also think returning the advertise_addr might also be the correct behaviour if it has been overridden.

I also think this would help with people understanding that you should only ever register and modify a service with the local agent - you can't do it with just any random agent in the cluster.

**Comment by slackpad** [comment]
Hey @moofish32 this looks like a good solution! I think we'd want to add a check to make sure there's nothing preceding `agent` and update the docs. It's probably fine to reuse `nodeLookup` though this technically doesn't need to go up to the Consul servers at all. That might be an optimization worth considering.

**Comment by moofish32** [comment]
@slackpad -- sorry for the delay on my end here, I missed your comment nearly a month ago. 

I updated the documentation and actually started trying to attempt the optimization. I ran in to a lot of edge cases that was making testing difficult. I started by trying to use agent config struct to pull `BindAddr` but discovered that might be `0.0.0.0`, then it might IPv6. At that point I started thinking of ways to use system calls to obtain the address and ran into the Docker deviations. 

For now, I have updated the documentation to reflect the fact a DNS lookup is going to the server for this call because I thought the edge cases would be hard to document and hard to test. Instead I was considering trying to use a caching optimization. Effectively, we know the IP can't change without restarting the agent and one dns lookup should be ok. What do you think?

I also see you cut 0.8.0, let me know if you want this in another version, branch etc.

**Comment by slackpad** [comment]
Hi @moofish32 want to do a quick rebase on this one?

File: website/source/docs/agent/dns.html.markdown [hunk 1]
```diff
@@ -198,7 +198,17 @@ Both A and SRV records are supported. SRV records provide the port that a servic
 registered on, enabling clients to avoid relying on well-known ports. SRV records are
 only served if the client specifically requests them.
 
-### UDP Based DNS Queries
+## Agent Query Lookup
+
+The agent query is designed to return IP information about the local Consul 
+agent. The IP information returned is the DNS answer for the local node name.
+This does in fact send a DNS query from the local instance to the Consul Server
+Quorum. The result is the IP informaiton that any other service would receive
```
Comment (spawluk) [inline website/source/docs/agent/dns.html.markdown]: Typo: Information

**Comment by whit537** [comment]
> want to do a quick rebase on this one?

Done in #3805. Suggest closing this in favor of that.

File: website/source/docs/agent/dns.html.markdown (hunk 1 above)
Comment (whit537) [reply website/source/docs/agent/dns.html.markdown]: Done in 65f112ae5f487d6f8574c2e731c5a199d0b60589 over in #3805.

**Comment by moofish32** [comment]
@whit537  @slackpad -- too many emails and I didn't see this pop back up. I rebased and updated the tests, happy to have you guys move this to another PR. Hopefully this fixes the tests. I grabbed the typo as well from #3805.

**Comment by moofish32** [comment]
seeing intermittent failures on this: 

`testagent.go:197: TestPreparedQuery_Execute/#00Catalog.ListNodes failed:No cluster leader`

Not sure if that's from this PR though?

**Comment by whit537** [comment]
> seeing intermittent failures on this

A chronic problem, cf. https://github.com/hashicorp/consul/pull/3174#issuecomment-357792856.

**Comment by whit537** [comment]
> A chronic problem

... deserving of its own ticket: https://github.com/hashicorp/consul/issues/3807. ☺️

**Comment by i0rek** [comment]
Hi! I hope you don't mind that I rebased your branch so that it is up to date. I will help you get this feature merged, unless I am missing something this seems like a nice improvement.

**Comment by i0rek** [comment]
After giving it more thought I discovered a couple of things:

* is this still needed or are there better solutions at this point
* this is probably not enough because at that point you might as well want to discover services and ports on your local agent, because otherwise you would have to hardcode them

Which is why I am closing this PR now. I would be happy to reopen though in case somebody still cares about it and didn't find another solution.

Thanks again!
